can't print the vertical-bars operator

```
module Issue_179_cant_print_vertbars;

type Mbrs = unit {
 : bytes &size = 1;
};
 
type showItFail = unit {
 members : (Mbrs)[];
 on %done {
    print (|self.members|);
 }
};
```
reports
```
2126 /usr/local/etc/tmp/spicy$printf "1\n" | build/bin/spicy-driver  ../Issue_179_print_vertbar.spicy 
In file included from Issue_179_cant_print_vertbars.cc:6:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/libhilti.h:5:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/configuration.h:8:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/util.h:30:
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:31:12: error: call to 'to_string' is ambiguous
    return to_string(x, detail::adl::tag{});
           ^~~~~~~~~
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:69:23: note: in instantiation of function template specialization 'hilti::rt::to_string<unsigned long>' requested here
    return hilti::rt::to_string(x);
                      ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:81:20: note: in instantiation of function template specialization 'hilti::rt::detail::to_string_for_print<unsigned long>' requested here
    return detail::to_string_for_print(x);
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/hilti.h:20:29: note: in instantiation of function template specialization 'hilti::rt::to_string_for_print<unsigned long>' requested here
    std::cout << hilti::rt::to_string_for_print(t);
                            ^
Issue_179_cant_print_vertbars.cc:168:159: note: in instantiation of function template specialization 'hilti::rt::print<unsigned long>' requested here
extern void __hlt::Issue_179_cant_print_vertbars::__hook_showItFail___on_0x25_done_0x7ff8a3c39500(hilti::rt::ValueReference<showItFail>& __self) { hilti::rt::print(hilti::rt::optional::value((*__self).members, "../Issue_179_print_vertbar.spicy:10").size(), true); }
                                                                                                                                                              ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/bool.h:10:20: note: candidate function
inline std::string to_string(bool x, adl::tag /*unused*/) { return (x ? "True" : "False"); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:51:20: note: candidate function
inline std::string to_string(uint64_t x, adl::tag /*unused*/) { return fmt("%" PRIu64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:53:20: note: candidate function
inline std::string to_string(int64_t x, adl::tag /*unused*/) { return fmt("%" PRId64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:55:20: note: candidate function
inline std::string to_string(uint32_t x, adl::tag /*unused*/) { return fmt("%" PRIu32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:57:20: note: candidate function
inline std::string to_string(int32_t x, adl::tag /*unused*/) { return fmt("%" PRId32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:59:20: note: candidate function
inline std::string to_string(uint16_t x, adl::tag /*unused*/) { return fmt("%" PRIu16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:61:20: note: candidate function
inline std::string to_string(int16_t x, adl::tag /*unused*/) { return fmt("%" PRId16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:63:20: note: candidate function
inline std::string to_string(uint8_t x, adl::tag /*unused*/) { return fmt("%" PRIu8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:65:20: note: candidate function
inline std::string to_string(int8_t x, adl::tag /*unused*/) { return fmt("%" PRId8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/real.h:25:20: note: candidate function
inline std::string to_string(double x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:19:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:23:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:27:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:31:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:35:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:39:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:43:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint8_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:47:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int8_t> x, adl::tag /*unused*/) {
                   ^
1 error generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit Issue_179_cant_print_vertbars to bitcode
[error] spicy-driver: aborting after errors
```

--- a/hilti/CMakeLists.txt
+++ b/hilti/CMakeLists.txt
@@ -270,7 +270,8 @@ add_executable(hilti-rt-tests
                src/rt/tests/fiber.cc
                src/rt/tests/reference.cc
                src/rt/tests/result.cc
-               src/rt/tests/stream.cc)
+               src/rt/tests/stream.cc
+               src/rt/tests/to_string.cc)
 target_compile_options(hilti-rt-tests PRIVATE "-Wall")
 target_link_libraries(hilti-rt-tests PRIVATE hilti-rt doctest)
 add_test(NAME hilti-rt-tests COMMAND ${CMAKE_BINARY_DIR}/bin/hilti-rt-tests)

--- a/hilti/include/rt/types/integer.h
+++ b/hilti/include/rt/types/integer.h
@@ -48,6 +48,8 @@ inline std::string to_string(hilti::rt::integer::safe<int8_t> x, adl::tag /*unus
     return fmt("%" PRId8, *x.Ptr());
 }
 
+inline std::string to_string(hilti::rt::integer::safe<char> x, adl::tag /*unused*/) { return fmt("%" PRId8, *x.Ptr()); }
+
 inline std::string to_string(uint64_t x, adl::tag /*unused*/) { return fmt("%" PRIu64, x); }
 
 inline std::string to_string(int64_t x, adl::tag /*unused*/) { return fmt("%" PRId64, x); }

--- a/hilti/include/rt/types/string.h
+++ b/hilti/include/rt/types/string.h
@@ -33,6 +33,11 @@ inline std::string to_string(std::string_view x, adl::tag /*unused*/) {
     return fmt("\"%s\"", escapeUTF8(x, true, true, true));
 }
 
+template<typename CharT, size_t N>
+inline std::string to_string(const CharT (&x)[N], adl::tag /*unused*/) {
+    return fmt("\"%s\"", escapeUTF8(x, true, true, true));
+}
+
 } // namespace detail::adl
 
 template<>

--- a/hilti/include/rt/types/vector.h
+++ b/hilti/include/rt/types/vector.h
@@ -224,9 +224,15 @@ inline bool operator!=(const Empty& /*unused*/, const Vector<T, Allocator>& v) {
 namespace detail::adl {
 template<typename T, typename Allocator>
 inline std::string to_string(const Vector<T, Allocator>& x, adl::tag /*unused*/) {
+    using detail::adl::to_string;
     return fmt("[%s]",
                rt::join(rt::transform(x, [](const std::optional<T>& y) { return (y ? rt::to_string(*y) : "(unset)"); }),
                         ", "));
+}
+
+template<typename T, typename Allocator>
+inline std::string to_string(const std::vector<T, Allocator>& x, adl::tag /*unused*/) {
+    return to_string(static_cast<const Vector<T, Allocator>&>(x), adl::tag{});
 }
 
 template<typename T, typename Allocator>

--- a/hilti/src/rt/tests/to_string.cc
+++ b/hilti/src/rt/tests/to_string.cc
@@ -1,0 +1,42 @@
+// Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
+
+#include <doctest/doctest.h>
+
+#include <cstdint>
+
+#include <hilti/rt/extension-points.h>
+#include <hilti/rt/hilti.h>
+#include <hilti/rt/libhilti.h>
+#include <hilti/rt/types/optional.h>
+#include <hilti/rt/types/vector.h>
+
+using namespace hilti::rt;
+
+TEST_CASE("to_string") {
+    SUBCASE("primitive") {
+        CHECK_EQ(to_string(true), "True");
+        CHECK_EQ(to_string(false), "False");
+        CHECK_EQ(to_string(-1), "-1");
+        CHECK_EQ(to_string(0), "0");
+        CHECK_EQ(to_string(1), "1");
+        CHECK_EQ(to_string(2), "2");
+        CHECK_EQ(to_string(1.5), "1.5");
+        CHECK_EQ(to_string(1.5), "1.5");
+    }
+
+    SUBCASE("string") { CHECK_EQ(to_string(std::string("abc")), "\"abc\""); }
+
+    SUBCASE("vector") {
+        CHECK_EQ(to_string(std::vector<int8_t>()), "[]");
+        CHECK_EQ(to_string(std::vector{int8_t{1}}), "[1]");
+        CHECK_EQ(to_string(std::vector{int8_t{1}, 2}), "[1, 2]");
+        CHECK_EQ(to_string(std::vector{std::vector{int8_t{1}, 2}}), "[1, 2]");
+    }
+
+    SUBCASE("optional") {
+        CHECK_EQ(to_string(std::optional<int8_t>()), "(not set)");
+        CHECK_EQ(to_string(std::optional<int8_t>(2)), "2");
+        CHECK_EQ(to_string(std::optional<std::optional<int8_t>>()), "(not set)");
+        CHECK_EQ(to_string(std::optional<std::optional<int8_t>>(2)), "2");
+    }
+}

--- a/hilti/src/rt/tests/to_string.cc
+++ b/hilti/src/rt/tests/to_string.cc
@@ -22,6 +22,7 @@ TEST_CASE("to_string") {
         CHECK_EQ(to_string(2), "2");
         CHECK_EQ(to_string(1.5), "1.5");
         CHECK_EQ(to_string(1.5), "1.5");
+        CHECK_EQ(to_string("abc"), "\"abc\"");
     }
 
     SUBCASE("string") { CHECK_EQ(to_string(std::string("abc")), "\"abc\""); }

--- a/hilti/src/rt/tests/to_string.cc
+++ b/hilti/src/rt/tests/to_string.cc
@@ -4,11 +4,7 @@
 
 #include <cstdint>
 
-#include <hilti/rt/extension-points.h>
-#include <hilti/rt/hilti.h>
 #include <hilti/rt/libhilti.h>
-#include <hilti/rt/types/optional.h>
-#include <hilti/rt/types/vector.h>
 
 using namespace hilti::rt;
 
@@ -23,6 +19,12 @@ TEST_CASE("to_string") {
         CHECK_EQ(to_string(1.5), "1.5");
         CHECK_EQ(to_string(1.5), "1.5");
         CHECK_EQ(to_string("abc"), "\"abc\"");
+    }
+
+    SUBCASE("safe-int") {
+        using integer::safe;
+        CHECK_EQ(to_string(safe<int8_t>(-1)), "-1");
+        CHECK_EQ(to_string(safe<char>(-1)), "-1");
     }
 
     SUBCASE("string") { CHECK_EQ(to_string(std::string("abc")), "\"abc\""); }

--- a/spicy/src/compiler/codegen/parser-builder.cc
+++ b/spicy/src/compiler/codegen/parser-builder.cc
@@ -337,7 +337,7 @@ struct ProductionVisitor
                             build_parse_stage1_logic();
 
                         auto result = build_parse_stage2_logic();
-                        builder()->addReturn(std::move(result));
+                        builder()->addReturn(result);
 
                         end_try(try_);
                         popState();
@@ -931,7 +931,7 @@ hilti::type::Struct ParserBuilder::addParserMethods(hilti::type::Struct s, const
             visitor.pushDestination(builder::id("unit"));
             visitor.parseProduction(*grammar.root());
             visitor.popDestination();
-            builder()->addReturn(builder::move(state().cur));
+            builder()->addReturn(state().cur);
             popState();
 
             auto body_ext_overload1 = popBuilder();
@@ -957,7 +957,7 @@ hilti::type::Struct ParserBuilder::addParserMethods(hilti::type::Struct s, const
         visitor.pushDestination(builder::id("unit"));
         visitor.parseProduction(*grammar.root());
         visitor.popDestination();
-        builder()->addReturn(builder::move(state().cur));
+        builder()->addReturn(state().cur);
         popState();
 
         auto body_ext_overload2 = popBuilder();

